lib/Dialect/FIRRTL/Transforms/InferReadWrite.cpp

```diff
@@ -0,0 +1,268 @@
+        auto readData = builder.create<SubfieldOp>(rwPort, "rdata");
+        // wmode is high when the port is in write mode. That is this can be
+        // connected to the write enable.
+        auto wmode = builder.create<SubfieldOp>(rwPort, "wmode");
+        auto writeData = builder.create<SubfieldOp>(rwPort, "wdata");
+        auto mask = builder.create<SubfieldOp>(rwPort, "wmask");
+        // Temp wires to replace the original memory connects.
+        auto rAddr = builder.create<WireOp>(addr.getType(), "readAddr");
+        auto wAddr = builder.create<WireOp>(addr.getType(), "writeAddr");
+        auto wEnWire = builder.create<WireOp>(enb.getType(), "writeEnable");
+        auto rEnWire = builder.create<WireOp>(enb.getType(), "readEnable");
+        // addr = Mux(WriteEnable, WriteAddress, ReadAddress).
+        builder.create<ConnectOp>(
+            addr, builder.create<MuxPrimOp>(wEnWire, wAddr, rAddr));
+        // Enable = Or(WriteEnable, ReadEnable).
+        builder.create<ConnectOp>(enb,
+                                  builder.create<OrPrimOp>(rEnWire, wEnWire));
+        // WriteMode = WriteEnable.
+        builder.create<ConnectOp>(wmode, wEnWire);
+        // Now iterate over the original memory read and write ports.
+        for (auto portIt : llvm::enumerate(mem.results())) {
+          // Get the port value.
+          Value portVal = portIt.value();
+          // Get the port kind.
+          bool readPort =
+              mem.getPortKind(portIt.index()) == MemOp::PortKind::Read;
+          // Iterate over all users of the port, which are the subfield ops, and
+          // replace them.
+          for (Operation *u : portVal.getUsers())
+            if (auto sf = dyn_cast<SubfieldOp>(u)) {
+              auto fName =
+                  sf.input().getType().cast<BundleType>().getElementName(
+                      sf.fieldIndex());
+              if (fName.equals("en")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(rEnWire);
+                else
+                  sf->replaceAllUsesWith(wEnWire);
+              } else if (fName.equals("clk")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(clk);
+                else {
+                  auto dummyWire = builder.create<WireOp>(sf.getType());
+                  sf->replaceAllUsesWith(dummyWire);
+                }
+              } else if (fName.equals("addr")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(rAddr);
+                else
+                  sf->replaceAllUsesWith(wAddr);
+              } else if (fName.equals("data")) {
+                if (readPort)
+                  sf->replaceAllUsesWith(readData);
+                else
+                  sf->replaceAllUsesWith(writeData);
+              } else if (fName.equals("mask")) {
+                if (!readPort)
+                  sf->replaceAllUsesWith(mask);
+              }
```

Comment:
You may be able to use a `StringSwitch` here. That has some restrictions in that it has to return something, but I think you could return the replacement value.

Reply:
Also, this may be clearer as `if(readPort)`/`else` as opposed to the boolean with almost every port doing something different.

Reply:
Thanks, simplified it using the `StringSwitch`